Fix SceneView UI layout issues with 2021.2+

Kerfuffles opened the pull request:
This PR should resolve #327, and should self-adjust to any docked overlay toolbars automatically. It also fixes the scene info overlay in the same way, as it also relied on `SceneView.postion` to get the width/height of the scene view. The new method should use `SceneView.rootVisualElement.contentRect` instead, which accounts for any docked overlays, etc.. This PR does NOT fix the issue of overlays blocking the RCSG windows/toolbar, nor does it fix #318

Kerfuffles commented:
In the future, it may be worthwhile to re-visit this and provide an actual proper fix (actually use overlays), so that there are no overlapping windows or other problems going forward. It *should* theoretically be possible to just call the IMGUI code inside of an overlay, instead of rewriting the entire thing. Just some thoughts.

LogicalError commented on RealtimeCSG/Assets/Plugins/RealtimeCSG/Editor/Scripts/Control/Managers/SceneViewEventHandler.cs:
```diff
@@ -70,8 +70,8 @@ internal static void OnScene(SceneView sceneView)
-					SceneViewBottomBarGUI.ShowGUI(sceneView);
-                	SceneViewInfoGUI.DrawInfoGUI( sceneView );
+					//SceneViewBottomBarGUI.ShowGUI(sceneView);
+                	//SceneViewInfoGUI.DrawInfoGUI( sceneView );
```
I vaguely remember this was a Unity bug workaround .. so might need an ifdef around it (if it's been fixed in unity)

Kerfuffles replied:
do you know what version this may have been introduced in?

LogicalError replied:
I'm guessing the version that introduced the new sceneview UI stuff might've changed things, so might be better to keep things the way they were for earlier unity versions, and switch to the new things for versions after

Kerfuffles replied:
yeah. i haven't noticed any issues cropping up from this. keeping those two lines actually causes the toolbar to be drawn twice. once in the correct place, and again above it.

LogicalError commented on RealtimeCSG/Assets/Plugins/RealtimeCSG/Editor/Scripts/Control/Managers/UpdateLoop.cs:
```diff
@@ -64,9 +64,10 @@ void Initialize()
 #endif
 
-#if UNITY_2018_3_OR_NEWER
+#if UNITY_2018_3_OR_NEWER && !UNITY_2020_1_OR_NEWER
```
what about < 2018.3 ?

Kerfuffles replied:
this is how it was written already, so im not sure what you mean?

Kerfuffles replied:
i realized what you had meant with this, and it has been fixed with the latest commit

LogicalError commented on RealtimeCSG/Assets/Plugins/RealtimeCSG/Editor/Scripts/View/GUI/EditModeGUI/EditModeManager.cs:
```diff
@@ -517,7 +517,7 @@ public static void OnSceneGUI(SceneView sceneView)
 					{
 						// handle the tool
-						var sceneSize = sceneView.position.size;
+						var sceneSize = sceneView.rootVisualElement.contentRect;
```
you might want to guard this with an ifdef for users of older unity versions, not sure when this was introduced ..

Kerfuffles replied:
this specific part was reverted back to master as it caused breaking changes as was mentioned by @HeadClot and @andreiagmu.

HeadClot commented:
Hey, I am running into an issue with this PR. I cannot draw out any form of CSG Shapes. Is anyone else running into this issue?

I am on Unity 2021.2.7f1 and I am getting no errors.

andreiagmu commented:
@Kerfuffles @LogicalError I'm having the same issue with this PR that @HeadClot commented above.
I'm testing in Unity 2021.2.6f1, and I can't draw any CSG Shape in the viewport.
In the Master branch, the shapes drawing works correctly.

andreiagmu commented:
I fixed the issue, line 520 from EditModeManager.cs should be
`var sceneSize = sceneView.position.size;`
like in the master branch.

Now RCSG is working properly on Unity 2021.2 here :)

Kerfuffles commented:
> I fixed the issue, line 520 from EditModeManager.cs should be `var sceneSize = sceneView.position.size;` like in the master branch.
> 
> Now RCSG is working properly on Unity 2021.2 here :)

This did indeed fix the issue! Thanks for that! In addition to those changes, I changed the sliders in the sphere and cylinder generators, to buttons, which is a change I had done in my own personal project as a workaround to the buggy input with sliders in the scene view.

HeadClot commented:
Cannot wait to see this merged :)

LogicalError commented:
Actually, I'm not sure if it's a good idea to commit all the ProjectSettings assets + ProjectVersion.txt since that will cause problems with older versions of unity

LogicalError commented on RealtimeCSG/Assets/Plugins/RealtimeCSG/Editor/Scripts/View/GUI/EditModeGUI/EditModeToolWindow.Editor.cs:
```diff
@@ -10,10 +10,7 @@ public class EditModeToolWindowEditor : Editor
-		{
-			Selection.activeObject = null;
 			return;
-		}
```
This shouldn't have been removed, this was a fix

Kerfuffles replied:
i thought i had changed that back. maybe i forgot to push the changes

Kerfuffles replied:
on checking this, i did indeed change it back, but git did not find the changes. should be fixed now

LogicalError commented on RealtimeCSG/Assets/Plugins/RealtimeCSG/Editor/Scripts/View/GUI/SceneViewBottomBarGUI/SceneViewBottomBar.GUI.cs:
```diff
@@ -24,13 +24,8 @@ public static void ShowGUI(SceneView sceneView, bool haveOffset = true)
-#else
 					bottomBarRect = new Rect(0, height - (CSG_GUIStyleUtility.BottomToolBarHeight + 18), 
 											  width, CSG_GUIStyleUtility.BottomToolBarHeight);
-#endif
```
This was part of the other PR, are you sure this should've been removed?

Kerfuffles replied:
yes. its not needed with changes from this PR, and does basically nothing as a result

LogicalError commented on RealtimeCSG/Assets/Plugins/RealtimeCSG/Editor/Scripts/View/GUI/SceneViewBottomBarGUI/SceneViewBottomBar.GUI.cs:
```diff
@@ -24,13 +24,8 @@ public static void ShowGUI(SceneView sceneView, bool haveOffset = true)
-#else
 					bottomBarRect = new Rect(0, height - (CSG_GUIStyleUtility.BottomToolBarHeight + 18), 
 											  width, CSG_GUIStyleUtility.BottomToolBarHeight);
-#endif
```
Again, are you sure this should've been removed?

Kerfuffles commented:
> Actually, I'm not sure if it's a good idea to commit all the ProjectSettings assets + ProjectVersion.txt since that will cause problems with older versions of unity

I would *like* to change the repo to a package, or to just the `Plugins/` folder. There seems to be a common confusion issue from others due to the entire repo being a project.

LogicalError commented:
> > Actually, I'm not sure if it's a good idea to commit all the ProjectSettings assets + ProjectVersion.txt since that will cause problems with older versions of unity
> 
> I would _like_ to change the repo to a package, or to just the `Plugins/` folder. There seems to be a common confusion issue from others due to the entire repo being a project.

Fair enough, but addressing those issues would make more sense as a separate PR since we'd need to figure out how to best do this

Kerfuffles commented:
> > > Actually, I'm not sure if it's a good idea to commit all the ProjectSettings assets + ProjectVersion.txt since that will cause problems with older versions of unity
> > 
> > 
> > I would _like_ to change the repo to a package, or to just the `Plugins/` folder. There seems to be a common confusion issue from others due to the entire repo being a project.
> 
> Fair enough, but addressing those issues would make more sense as a separate PR since we'd need to figure out how to best do this

Right. That was the plan. I may look into doing that at a later time, though.

Kerfuffles commented:
I am basically done changing anything with this PR, if nothing else needs fixing/revision.

LogicalError commented on RealtimeCSG/ProjectSettings/ProjectVersion.txt:
```diff
@@ -1 +1,2 @@
-m_EditorVersion: 5.6.4f1
+m_EditorVersion: 2021.2.7f1
+m_EditorVersionWithRevision: 2021.2.7f1 (6bd9e232123f)
```
Can you revert this?
I'm fairly sure there are people who are using realtime-csg who are on older versions than 2021
upgrading is easier than downgrading.
Better yet, just delete the file

LogicalError replied:
This would be the last issue

Kerfuffles replied:
Done.
